In [22]:
import json
import pandas as pd

In [23]:
data = pd.read_json('sizes.json', lines=True)


In [24]:
data.head()

__v                                   _id  \
0    0  {'$oid': '557abe8565950003006fefe7'}   
1    0  {'$oid': '557abe8565950003006fefe5'}   
2    0  {'$oid': '557abe8565950003006fefe8'}   
3    0  {'$oid': '557abe8565950003006fefe6'}   
4    0  {'$oid': '557abe8565950003006fefe9'}   

                              created_at filter_label label  old_label  \
0  {'$date': '2015-06-12T11:12:05.248Z'}            S    26      [26\]   
1  {'$date': '2015-06-12T11:12:05.244Z'}           XS    24  [24, 24\]   
2  {'$date': '2015-06-12T11:12:05.249Z'}            S    28      [28\]   
3  {'$date': '2015-06-12T11:12:05.247Z'}            S    25       [25]   
4  {'$date': '2015-06-12T11:12:05.249Z'}            M    30      [30\]   

                              updated_at  
0  {'$date': '2015-07-07T16:29:11.868Z'}  
1  {'$date': '2015-07-07T16:27:03.743Z'}  
2  {'$date': '2015-07-07T16:36:22.139Z'}  
3  {'$date': '2015-07-07T16:29:03.370Z'}  
4  {'$date': '2015-07-07T16:44:25.382Z'}

In [25]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51210 entries, 0 to 51209
Data columns (total 7 columns):
__v             51210 non-null int64
_id             51210 non-null object
created_at      51210 non-null object
filter_label    294 non-null object
label           51210 non-null object
old_label       51210 non-null object
updated_at      51210 non-null object
dtypes: int64(1), object(6)
memory usage: 2.7+ MB


**Observations**
- dataframe has 7 columns/ 51210 lines
- filter_label column has only 294 entries, vs 51 210 lines
- dataframe entries are objects which is not the most suitable type for data manipulation

**Data frame columns - detailed view**

In [26]:
column_names=list(data)

In [27]:
column_types=[type(data._id[0]),
type(data.__v[0]),
type(data.created_at[0]),
type(data.filter_label[0]),
type(data.label[0]),
type(data.old_label[0]),
type(data.updated_at[0])]

In [28]:
for i in range (0,len(column_names)):
    print(column_names[i],column_types[i])

__v <class 'dict'>
_id <class 'numpy.int64'>
created_at <class 'dict'>
filter_label <class 'str'>
label <class 'str'>
old_label <class 'list'>
updated_at <class 'dict'>


** Step 1 - flatten the dataframe** 

1.1 Column **_if**: explore dictionary keys and values

In [29]:
col_name='_id'

In [30]:
def flatten_dic(df,col_name):
    keys=[]
    values=[]
    for index, row in data.iterrows():
        item=row[col_name]
        keys.extend(item.keys())
        values.extend(item.values())
    return keys, values

In [31]:
keys_id, values_id= flatten_dic(data, col_name)

In [32]:
def check_unique(item):
    unq_items=set(item)
    return len(unq_items)

In [33]:
print(check_unique(keys_id))

1


In [34]:
print(check_unique(values_id))

51210


**Observations**
- The keys are the same across all lines, therefore, can be removed, and only ID value can remain as item ID identifier

In [35]:
def clean_col(data,col_name):
    for index,row in data.iterrows():
        item=row[col_name]
        value=list(item.values())    
        data.loc[data.index[index], col_name]=str(value[0])
    return data
 


In [37]:
clean_col(data, '_id')

AttributeError: 'str' object has no attribute 'values'

In [39]:
data.head()

__v                       _id                             created_at  \
0    0  557abe8565950003006fefe7  {'$date': '2015-06-12T11:12:05.248Z'}   
1    0  557abe8565950003006fefe5  {'$date': '2015-06-12T11:12:05.244Z'}   
2    0  557abe8565950003006fefe8  {'$date': '2015-06-12T11:12:05.249Z'}   
3    0  557abe8565950003006fefe6  {'$date': '2015-06-12T11:12:05.247Z'}   
4    0  557abe8565950003006fefe9  {'$date': '2015-06-12T11:12:05.249Z'}   

  filter_label label  old_label                             updated_at  
0            S    26      [26\]  {'$date': '2015-07-07T16:29:11.868Z'}  
1           XS    24  [24, 24\]  {'$date': '2015-07-07T16:27:03.743Z'}  
2            S    28      [28\]  {'$date': '2015-07-07T16:36:22.139Z'}  
3            S    25       [25]  {'$date': '2015-07-07T16:29:03.370Z'}  
4            M    30      [30\]  {'$date': '2015-07-07T16:44:25.382Z'}

In [41]:
col_name2='created_at'

In [42]:
keys_cr_at, values_cr_at=flatten_dic(data,col_name2)

In [43]:
print(check_unique(keys_cr_at))

1


In [44]:
print(check_unique(values_cr_at))

43535


In [45]:
### observation: there are 43 535 craeted_at values, less than lines in the DF, therefore, we can assume this column has nan values 

In [47]:
clean_col(data, col_name2)

AttributeError: 'str' object has no attribute 'values'

In [48]:
data.head()

__v                       _id                created_at filter_label label  \
0    0  557abe8565950003006fefe7  2015-06-12T11:12:05.248Z            S    26   
1    0  557abe8565950003006fefe5  2015-06-12T11:12:05.244Z           XS    24   
2    0  557abe8565950003006fefe8  2015-06-12T11:12:05.249Z            S    28   
3    0  557abe8565950003006fefe6  2015-06-12T11:12:05.247Z            S    25   
4    0  557abe8565950003006fefe9  2015-06-12T11:12:05.249Z            M    30   

   old_label                             updated_at  
0      [26\]  {'$date': '2015-07-07T16:29:11.868Z'}  
1  [24, 24\]  {'$date': '2015-07-07T16:27:03.743Z'}  
2      [28\]  {'$date': '2015-07-07T16:36:22.139Z'}  
3       [25]  {'$date': '2015-07-07T16:29:03.370Z'}  
4      [30\]  {'$date': '2015-07-07T16:44:25.382Z'}

In [49]:
data['created_at']  = pd.to_datetime(data['created_at'])

In [61]:
data.head()

__v                       _id              created_at filter_label label  \
0    0  557abe8565950003006fefe7 2015-06-12 11:12:05.248            S    26   
1    0  557abe8565950003006fefe5 2015-06-12 11:12:05.244           XS    24   
2    0  557abe8565950003006fefe8 2015-06-12 11:12:05.249            S    28   
3    0  557abe8565950003006fefe6 2015-06-12 11:12:05.247            S    25   
4    0  557abe8565950003006fefe9 2015-06-12 11:12:05.249            M    30   

   old_label              updated_at  
0      [26\] 2015-07-07 16:29:11.868  
1  [24, 24\] 2015-07-07 16:27:03.743  
2      [28\] 2015-07-07 16:36:22.139  
3       [25] 2015-07-07 16:29:03.370  
4      [30\] 2015-07-07 16:44:25.382

In [62]:
col_name3='updated_at'

In [63]:
keys_updt_at, values_updt_at=flatten_dic(data,col_name3)

AttributeError: 'Timestamp' object has no attribute 'keys'

In [54]:
print(check_unique(keys_updt_at))

1


In [55]:
print(check_unique(values_updt_at))

46594


In [56]:
### observation: there are 46 594 craeted_at values, less than lines in the DF, therefore, we can assume this column has nan values 

In [ ]:
clean_col(data, col_name3)

In [58]:
data['updated_at']  = pd.to_datetime(data['updated_at'])

In [64]:
data.head()

__v                       _id              created_at filter_label label  \
0    0  557abe8565950003006fefe7 2015-06-12 11:12:05.248            S    26   
1    0  557abe8565950003006fefe5 2015-06-12 11:12:05.244           XS    24   
2    0  557abe8565950003006fefe8 2015-06-12 11:12:05.249            S    28   
3    0  557abe8565950003006fefe6 2015-06-12 11:12:05.247            S    25   
4    0  557abe8565950003006fefe9 2015-06-12 11:12:05.249            M    30   

   old_label              updated_at  
0      [26\] 2015-07-07 16:29:11.868  
1  [24, 24\] 2015-07-07 16:27:03.743  
2      [28\] 2015-07-07 16:36:22.139  
3       [25] 2015-07-07 16:29:03.370  
4      [30\] 2015-07-07 16:44:25.382

In [67]:
data[['old_label','filter_label','label']]

old_label filter_label  \
0                                                  [26\]            S   
1                                              [24, 24\]           XS   
2                                                  [28\]            S   
3                                                   [25]            S   
4                                                  [30\]            M   
5      [6 - 12m, 6-12 months, 6-12m, 6- 12 Months, 6 ...    0-1 years   
6      [3 - 6 months, 3 - 6m, 3-6 Months, 3-6m, EU 16...    0-1 years   
7      [2yrs, 2 years, 2y, Age 2, 2 years (92), 2 yea...    1-2 years   
8                                                   [54]          NaN   
9                                                   [48]          NaN   
10                                        [12-24 months]    1-2 years   
11                                                   [C]          NaN   
12                                                [sand]          NaN   
13     [Pink, fluorescent pink, pink grapefruit, Pink...          NaN   
14                                                   [D]          NaN   
15                                          [Anthracite]          NaN   
16                      [N, UK N, Size N, 53, Large N-O]          NaN   
17                                       [Full alphabet]          NaN   
18     [12-18m, 12 - 18m, 12 - 18 Months, 12-18 month...    0-1 years   
19     [18-24m, 18 - 24m, 18 - 24 Months, 18-24, 18-2...    1-2 years   
20     [6 years, Age 6, 6y, 6 years (116), 6 years (1...    5-7 years   
21                                             [R, UK R]          NaN   
22     [1 - 2yrs, 1-2 years, 1- 2yrs, 1 - 2 years, 1-...    1-2 years   
23                                      , 12 Months]    0-1 years   
24     [7 - 8 years, 7 - 8yrs, 7-8 yrs, 7-8 years, 7-...   8-10 years   
25     [2-3 years, 2 - 3yrs, 2-3y, 2 - 3 Years, Age 2...    2-4 years   
26                                     [UK 4.5 / EUR 21]    1-2 years   
27                                                [Teal]          NaN   
28           [8 years, Age 8, 8y, 8 years (128), 8years]   8-10 years   
29       [6-9 months, 6 - 9 months, 6 - 9m, 6-9 m, 6-9m]    0-1 years   
...                                                  ...          ...   
51180                                                 []          NaN   
51181                                                 []          NaN   
51182                                                 []          NaN   
51183                                                 []          NaN   
51184                                                 []          NaN   
51185                                                 []          NaN   
51186                                                 []          NaN   
51187                                                 []          NaN   
51188                                                 []          NaN   
51189                                                 []          NaN   
51190                                                 []          NaN   
51191                                                 []          NaN   
51192                                                 []          NaN   
51193                                                 []          NaN   
51194                                                 []          NaN   
51195                                                 []          NaN   
51196                                                 []          NaN   
51197                                                 []          NaN   
51198                                                 []          NaN   
51199                                                 []          NaN   
51200                                                 []          NaN   
51201                                                 []          NaN   
51202                                                 []          NaN   
51203                               